library/find_unused_disk_module.py

```diff
@@ -0,0 +1,108 @@
+#!/usr/bin/python
+
+DOCUMENTATION = '''
+---
+module: find_unused_disk_module
+short_description: Get first unused disk
+description:
+    - Disks are considered in ascending alphanumeric sorted orderself.
+    - First disk to meet all conditions is considered 'empty' and returned (using kernel device name).
+        - 1. No known signatures exist on the disk, with the exception of partition tables.
+        - 2. If there is a partition table on the disk, it contains no partitions.
+        - 3. The disk has no holders to eliminate the possibility of it being a multipath or dmraid member device.
+        - 4. Device can be opened with exclusive access to make sure no other software is using it.
+    - If no disks meet all criteria, "Unable to find empty disk" will be returned.
+author: Eda Zhou (@edamamez)
+'''
+
+EXAMPLES = '''
+- name: test finding first empty device module
+  hosts: localhost
+  tasks:
+    - name: run module
+      find_unused_disk_module:
+      register: testout
+    - name: dump test output
+      debug:
+        msg: '{{ testout }}'
+'''
+
+RETURN = '''
+disk_name:
+    description: Information about first empty disk
+    returned: On success
+    type: complex
+    samples: "/dev/sda"
+             "Unable to find empty disk"
+'''
+
+
+import os
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible.module_utils import facts
+
+
+# Ensure that no known signatures exist on the disk, with the exception of 
+# partition tables. (blkid -p <device>)
+def no_signature(run_command, disk):
+    signatures = run_command(['blkid', '-p', disk])
+    return not 'UUID' in signatures[1]
+
+
+# Ensure that the disk has no holders to eliminate the possibility of it being a
+#  multipath or dmraid member device. (/sys/class/block/<device>/holders/ empty)
+def no_holders(run_command, disk_name):
+    holders = run_command(['ls', '/sys/class/block/' + disk_name + '/holders/'])
+    return len(holders[1]) == 0
+
+
+# Ensure that you can open the device with exclusive access to make sure no 
+# other software is using it. (see man 2 open and look for O_EXCL)
+def can_open(open, close, disk):
+    try:
+        exclusive = open(disk, os.O_EXCL)
+        close(exclusive)
+        return True
+    except:
+        return False
+
+
+def run_module():
+    module_args = dict(
+        disk=dict(type='str', required=False)
+    )
+    
+    result = dict(
+        changed=False,
+        disk=''
+    )
+    
+    module = AnsibleModule(
+        argument_spec=module_args,
+        supports_check_mode=True
+    )
+
+    ansible_facts = facts.ansible_facts(module)
+    run_command = module.run_command
+    all_disks = run_command(['lsblk', '-o', 'kname', '-l', '-n', '-p'])
+    for disk in all_disks[1].split():
+        disk_name = os.path.basename(disk)
+        # Ensure that, if there is a partition table on the disk, it contains no partitions.    
+        try:
```

Comment:
A more typical way to do this would be something like this:

```python
if (not ansible_facts['devices'][disk_name]['partitions'] and
    no_signature(run_command, disk) and
    no_holders(run_command, disk_name) and
    can_open(os.open, os.close, disk)):
    result['disk'] = disk
    module.exit_json(**result)
```

Reply:
I believe I made it a try-except because some devices won't have a 'partitions' variable and the command ansible_facts['devices'][disk_name]['partitions'] will fail with some form of string not found exception.

Reply:
Bah, that isn't quite right. In the `try` you'd have to coerce that to a `bool`, so something like:

`has_partitions = (ansible_facts[...] == dict())` or something along those lines.

Reply:
Better yet, just do it explicitly: `has_partitions = bool(ansible_facts[...])`

Reply:
That's really smart, done!